File: typescript/src/tools/search_tools.ts
import { apiRequest } from "../client.js";
import { mcpError } from "../errors.js";
import type { RawToolsResponse } from "../schema.js";
import { searchToolsInputSchema, type SearchToolsInput } from "../types.js";

export const name = "toolrate_search_tools";

export const config = {
  title: "Search and browse rated tools in ToolRate",
  description: [
    "Search the ToolRate catalog by name, identifier substring, or category. Returns identifiers you can pass to toolrate_assess.",
    "",
    "Use this when the agent needs to find candidate tools for a task but doesn't know the canonical identifier.",
  ].join("\n"),
  inputSchema: searchToolsInputSchema,
};

export async function handler(input: SearchToolsInput) {
  try {
    const { body, rateLimit } = await apiRequest<RawToolsResponse>({
      method: "GET",
      path: "/v1/tools",
      query: {
        q: input.q,
        category: input.category,
        limit: input.limit,
        offset: input.offset,
      },
    });

    const result = {
      ...body,
      ...(rateLimit ? { _meta: { rate_limit: rateLimit } } : {}),
    };

    return {
      content: [{ type: "text" as const, text: JSON.stringify(result, null, 2) }],
    };
  } catch (err) {
    return mcpError(err);
  }
}

File: python/src/toolrate_mcp/tools/search_tools.py
"""``toolrate_search_tools`` — browse the ToolRate catalog."""
from __future__ import annotations

import json
from typing import Annotated, Any, Optional

from pydantic import Field

from ..client import api_request
from ..errors import raise_tool_error

NAME = "toolrate_search_tools"

DESCRIPTION = "\n".join(
    [
        "Search the ToolRate catalog by name, identifier substring, or category. Returns identifiers you can pass to toolrate_assess.",
        "",
        "Use this when the agent needs to find candidate tools for a task but doesn't know the canonical identifier.",
    ]
)


async def handler(
    q: Annotated[
        Optional[str],
        Field(default=None, max_length=256, description="Search by tool name or identifier (case-insensitive substring)."),
    ] = None,
    category: Annotated[Optional[str], Field(default=None, max_length=128, description="Filter by category.")] = None,
    limit: Annotated[Optional[int], Field(default=None, ge=1, le=200, description="Max results (default 50).")] = None,
    offset: Annotated[Optional[int], Field(default=None, ge=0, le=10_000, description="Pagination offset.")] = None,
) -> str:
    try:
        body, rate_limit = await api_request(
            method="GET",
            path="/v1/tools",
            query={"q": q, "category": category, "limit": limit, "offset": offset},
        )
    except Exception as e:
        raise_tool_error(e)

    result: dict[str, Any] = {**body}
    if rate_limit is not None:
        result["_meta"] = {"rate_limit": rate_limit}
    return json.dumps(result, indent=2)
